feat(regular_expression): implement visitor pattern trait for regex AST

camchenry opened the pull request:
- resolves https://github.com/oxc-project/oxc/issues/5977
- supersedes https://github.com/oxc-project/oxc/pull/5951

To facilitate easier traversal of the Regex AST, this PR defines a `Visit` trait with default implementations that will walk the entirety of the Regex AST. Methods in the `Visit` trait can be overridden with custom implementations to do things like analyzing only certain nodes in a regular expression, which will be useful for regex-related `oxc_linter` rules.

In the future, we should consider automatically generating this code as it is very repetitive, but for now a handwritten visitor is sufficient.

camchenry commented:
* **#6055** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6055?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @camchenry and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

leaysgur commented on crates/oxc_regular_expression/src/visit.rs:
```diff
@@ -0,0 +1,391 @@
+use walk::*;
+
+#[derive(Copy, Clone, Debug)]
+pub enum RegexAstKind<'a> {
```
IMO: I feel it might be better to avoid using the term `Regex`. It brings to mind the `regex` crate. 👀 

Although it's a bit verbose, how about `RegularExpressionAstKind`?

camchenry replied:
hm, I'm not strongly opposed to `RegularExpressionAstKind`, but it is kind of long. maybe to differentiate from the `regex` crate, we could use the term `RegExp` instead? so it would be `RegExpAstKind` instead. That is more similar to how the ECMA262 spec refers to it.

leaysgur replied:
Personally, I still prefer `RegularExpressionAstKind` (since it's the `AstKind` of `oxc_regular_expression`), but I think `RegExpAstKind` would also be fine. 👍🏻 

---

BTW, (feel free to ignore this if I'm wrong, 😓) is it possible to alias an export with the same name in Rust?

```rs
use oxc_ast::AstKind;
use oxc_regular_expression::AstKind as RegExpAstKind; // 👈🏻 
```

If it's possible, I’m starting to think that simply leaving it as `AstKind` might be a good option.

camchenry replied:
I think that's definitely possible, though it's less ideal for the linter codebase since it's pretty common to import AstKind too. Let's stick with this name for now and we can always change it later?

leaysgur replied:
> that's definitely possible

Thanks, I learned something new. 😃 

> Let's stick with this name for now

🙆🏻

leaysgur commented on crates/oxc_regular_expression/src/visit.rs:
```diff
@@ -0,0 +1,391 @@
+}
+
+pub mod walk {
+    use crate::ast::UnicodePropertyEscape;
```
Seems unnecessary?

leaysgur reviewed (approved):
Thank you! 🙌🏻

camchenry committed:
feat(regular_expression): implement visitor pattern trait for regex AST (#6055)

- resolves https://github.com/oxc-project/oxc/issues/5977
- supersedes https://github.com/oxc-project/oxc/pull/5951

To facilitate easier traversal of the Regex AST, this PR defines a `Visit` trait with default implementations that will walk the entirety of the Regex AST. Methods in the `Visit` trait can be overridden with custom implementations to do things like analyzing only certain nodes in a regular expression, which will be useful for regex-related `oxc_linter` rules.

In the future, we should consider automatically generating this code as it is very repetitive, but for now a handwritten visitor is sufficient.
